src/test/java/com/squareup/javapoet/AnnotatedTypeNameTest.java

```diff
@@ -47,6 +50,27 @@
     assertEquals(annotated, annotated.annotated());
   }
 
+  // https://github.com/square/javapoet/issues/431
+  @Test public void annotatedNestedType() {
+    String tuacanon = TypeUseAnnotation.class.getCanonicalName();
+    String expected = "java.util.Map.@" + tuacanon + " Entry";
```

Comment:
Have a look at https://github.com/square/javapoet/issues/431#issuecomment-190978668 - that's even more "wow!" and I didn't see this notation used in any Java source, yet.